Support for symlinks in mount / vfs

pasnox opened the pull request:
#### What is the purpose of this change?
Add mount / vfs symlink support

Please, note that it has only been tested with storj backend on linux.
Especially, macOS, windows were not tested.
Please do test those platforms and any other backends !!!

#### Was the change discussed in an issue or in the forum before?
https://github.com/rclone/rclone/issues/2975

#### Checklist
- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

#### Notes
The whole work has been (trashed mostly) refactored. They were flaws in the previous implementation, especially trying to use truncated names in the vfs dir internal nodes  names which lead to inconsistency when it comes to push changes to the remotes.
For this reason, the refactoring went in 2 parts:

1. Add symlinks support in the VFS itself, in the form of always storing symlinks with the .rclonelink suffix, those being regular files.
The VFS Readlink and Symlink additions allow support for symlinks, even if the --links command line switch is not enabled, it's just a mater of creating a .rclonelink regular file with the target inside.
The only difference is that when --link is enabled, the VFS nodes Mode() would report the **os.ModeSymlink** flag hint.
This approach limit adding regressions in the VFS itself as there is no more names tweaking.
Some more checks were also added, to avoid to allow creating / moving files / links / directories names that would clash.
By example, the VFS must not be able to create in the same directory  something called **Hello.txt** and something called **Hello.txt.rclonelink**, this is because when using the mount operation with --links enabled, we would end up with both files claiming the same node name: **Hello.txt**.

2.  Add symlinks support in the mount backends, they are required to translate from the full vfs names into truncated names to the end user: **Test.rclonelink** would be seen as **Test**.
When --links command line switch is enabled, the file system will translate any ***.rclonelink** files into * links.
On the other hand, if --links is disabled, the symlinks will be seen as regular files, with their .rclonelink suffix which content contains the target file path.

Those changes comes with a rather big TestSymlinks unit test, which test all possible api and combinations.
The tests are ran for a VFS and each backends, with and without --links command line switch enabled, covering all possible scenarios.

I did run the tests successfully using this bash script:
```
#!/bin/bash -e

# Run the VFS tests in all subdirectories
pushd vfs
go test -tags cmount -v ./...
#go test -tags cmount -v -run TestFunctional
popd

# This test uses the tests in vfs/vfstest
pushd cmd/cmount
go test -tags cmount -v
popd

# This test uses the tests in vfs/vfstest
pushd cmd/mount
go test -v
popd

# This test uses the tests in vfs/vfstest
pushd cmd/mount2
go test -v
popd
```

ncw commented on vfs/vfscommon/path.go:
```diff
@@ -5,6 +5,9 @@ import (
 )
 
+// Constants
+const LinkSuffix = ".rclonelink" // The suffix added to a translated symbolic link
```
This should probably go in `fs` as `vfs` is a higher layer above the backends and importing it from a backend looks weird!

ncw commented on cmd/mount2/node.go:
```diff
@@ -405,3 +406,27 @@ func (n *Node) Rename(ctx context.Context, oldName string, newParent fusefs.Inod
+var _ fusefs.NodeReadlinker = (*Node)(nil)
+
+// read symbolic link target
+func (n *Node) Readlink(ctx context.Context) (ret []byte, err syscall.Errno) {
```
Readlink should likely be a vfs function I think, `vfs.Readlink` analagous to `os.Readlink`

pasnox replied:
Done, though now symlink is created correctly (server wise) but local fs (fuse) thrown an error.
Any idea ??

```
 pasnox  ~  Projects  rclone  ./rclone mount storj-s3-pasnox:test /home/pasnox/public_html/storj-mount -v                                                                                                     SIGINT   master 
2022/07/29 20:00:40 INFO  : S3 bucket test: poll-interval is not supported by this remote
2022/07/29 20:00:49 INFO  : N3.rclonelink: Copied (new)
2022/07/29 20:00:49 fuse: panic in handler for Symlink [ID=0x6 Node=0x1 Uid=1000 Gid=1001 Pid=30309] from "N3" to target "Native-2": interface conversion: interface is nil, not fs.Node
goroutine 82 [running]:
bazil.org/fuse/fs.(*Server).serve.func2()
        /home/pasnox/go/pkg/mod/bazil.org/fuse@v0.0.0-20200524192727-fb710f7dfd05/fs/serve.go:994 +0x14e
panic({0x24cc400, 0xc000463560})
        /usr/lib/go/src/runtime/panic.go:838 +0x207
github.com/rclone/rclone/cmd/mount.(*Dir).Symlink(0xc0003320b0, {0x2c86860?, 0xc0003320b0?}, 0xc000c7e3c0)
        /home/pasnox/Projects/rclone/cmd/mount/dir.go:266 +0x3c5
bazil.org/fuse/fs.(*Server).handleRequest(0xc0007b8000, {0x2c98520, 0xc000a860c0}, {0x2c86860?, 0xc0003320b0}, 0xc0000644c0, {0x2c92a80?, 0xc000c7e3c0?}, 0xc0007bdf08)
        /home/pasnox/go/pkg/mod/bazil.org/fuse@v0.0.0-20200524192727-fb710f7dfd05/fs/serve.go:1101 +0x296a
bazil.org/fuse/fs.(*Server).serve(0xc0007b8000, {0x2c92a80, 0xc000c7e3c0})
        /home/pasnox/go/pkg/mod/bazil.org/fuse@v0.0.0-20200524192727-fb710f7dfd05/fs/serve.go:1015 +0x634
bazil.org/fuse/fs.(*Server).Serve.func1()
        /home/pasnox/go/pkg/mod/bazil.org/fuse@v0.0.0-20200524192727-fb710f7dfd05/fs/serve.go:512 +0x69
created by bazil.org/fuse/fs.(*Server).Serve
        /home/pasnox/go/pkg/mod/bazil.org/fuse@v0.0.0-20200524192727-fb710f7dfd05/fs/serve.go:510 +0x3be
```

pasnox replied:
I found out I would need to still fill / create the fuse node.
How would I test mount2/cmount on linux ?
Right now i'm just stucked using mount.

pasnox replied:
Nevermind.. find out myself.

ncw reviewed (commented):
I gave the code a very quick review - great work :-)

It needs tests though, both in the vfs layer and in the mount layer. The mount tests are in `vfs/vfstest/` and likely they will need some build constraints not to run on windows. Or maybe symlinks do work on windows - I don't know!

I don't see patches for `cmd/cmount`? This is the variant which runs on Windows/Mac by default. It also runs on Linux too. You'll need `-tags cmount` to build it.

pasnox commented:
> I gave the code a very quick review - great work :-)
> 
> It needs tests though, both in the vfs layer and in the mount layer. The mount tests are in `vfs/vfstest/` and likely they will need some build constraints not to run on windows. Or maybe symlinks do work on windows - I don't know!
> 
> I don't see patches for `cmd/cmount`? This is the variant which runs on Windows/Mac by default. It also runs on Linux too. You'll need `-tags cmount` to build it.

cmount support added.
Now all 3 mount variants do works correctly, as per tested by me on linux/storj.
Tests needs to be performed for others platforms/backends (especially local file is to be verified).
I'm a bit lazy to add tests (as I said, I don't know much Go, I just quickly read/learned it to fix this symlinks issue), but I will try to push some in the next days.

ncw commented:
Great work :-)

Let me know if you want help writing the tests.

pasnox commented:
> Let me know if you want help writing the tests.

Thanks, that would be greatly appreciated.

How do I build unit tests and run them ?
I saw there is plenty of places where linkSuffix was not updated correctly in tests, but I can't find a way to try to build/run them so that I can fix them all.

pasnox commented:
@ncw I just rebased atop master.
I'm trying to add tests, but I have to be honest, i'm lost.
How would I test mount (3 variants) and test Readlink / Symlink on those ?
Could you guide me or try to write skeleton ?
Af far as I can see symlinks should be supported on windows.

ncw commented:
There are probably two places that tests need to be added.

The first is in `vfs/*_test.go`

To pick an example

https://github.com/rclone/rclone/blob/1107da7247db32199fde92f42c916bcb4c60dc3f/vfs/dir_test.go#L500

You can see this creates a test VFS, creates files in it, then checks to see the files are as required. So you'd want to create a symlink and check it comes back in the listing.

You'll need to add tests to `vfs/dir_test.go` `vfs/file_test.go`, `vfs/vfs_test.go` and `vfs.write_test.go` to match the changes you made in each of those files.

Then you need to write a higher level test here https://github.com/rclone/rclone/tree/master/vfs/vfstest - the files in here are tested against the VFS itself and against all the different mount types. You write tests here just by creating and manipulating files - this is done either with vfs methods or os methods - the test runner abstracts that away.

That is quite a lot of tests I know, but it is the only way we keep such a complicated bit of software working. The VFS layer is by far the most complex bit of rclone and I often think its the bit of code I was only just clever enough to write and therefore it is too difficult for me to debug!

One thing I noted is that this should probably be gated on a flag - I don't think you have that at the moment?

pasnox commented:
Thanks for the hints, will have a look.

> One thing I noted is that this should probably be gated on a flag - I don't think you have that at the moment?
Right I do not have a flag for that. Will add it at some point in time ;)

pasnox commented:
Hm, how would I trigger tests ?

I tried `make quicktest` but it lead to plenty of failing tests, even with bare master ?

ncw commented:
You cd into the directory with the tests and run `go test -v` that will work for `vfs` `cmd/mount` etc

hongkongkiwi commented:
When can we get this merged?

pasnox commented:
Not yet, while it works fine cache less, it cause troubles with the cache system due to the names translations. Also still writing the unit tests to ensure correctness

pasnox commented:
@ncw I would like your inputs. I have refactored the current work into a new branch @ https://github.com/pasnox/rclone/commits/master-vfs-symlinks
The changes are much proper and enable a --link flags. I redesigned the changes so that the VFS do not expose translated names for symlinks so that the cache do not break. I only translate leaf names for the fuse list dir request. Now the issue is that it's quite difficult to translate names back for the VFS. See the new branch latest commit.
I have to make a double lookup to guess for symlinks, which is quite annoying. I did not found a proper way for that. Would you have an idea ? I though I could get back the Dirent type I fill in ReadDirAll, so that I cna then avoid the double lookup, but I did not found a way :(

pasnox commented:
@ncw Neverming, I found VFS.Dir.stat which is where I think node lookup should be tweaked, similarly to the insensitive lookup.
Refactoring is nearly finished, then I can write all those tests :)

pasnox commented:
@ncw Open for review now with unit tests and --links command line switch.

pasnox commented:
@ncw Hey, is anything missing so that a deep review being made ?

pasnox commented:
@albertony could you enable the workflow at least? I would like to ensure everything is ok and tests pass correctly before I lost the knowledge here

pasnox commented:
> @albertony could you enable the workflow at least? I would like to ensure everything is ok and tests pass correctly before I lost the knowledge here

Remaining CI issues seems unrelated to my changes ?

albertony commented:
> Remaining CI issues seems unrelated to my changes ?

To me it seems some of them are symlink/vfs/mount related and worth taking a closer look into..

pasnox commented:
> > Remaining CI issues seems unrelated to my changes ?
> 
> To me it seems some of them are symlink/vfs/mount related and worth taking a closer look into..

Should all be fixed now, thanks.

pasnox commented:
@ncw @albertony Is there a way to grab builds artifacts somewhere ? I would be now interested to live test symlinks on my target devices which are based on ARM 32 & 64bits.

albertony commented:
I don't think so. But, by pushing as a branch to upstream (github.com/rclone/rclone), binaries will be automatically uploaded to https://beta.rclone.org/branch/. I'm not sure if you will have permissions to do that, though, so one of us may have to do it..

pasnox commented:
> 

Bah, need to wait for somebody to review and have it merged then... thanks.

albertony commented:
I pushed a branch [pasnox-symlink](https://github.com/rclone/rclone/tree/pasnox-symlink) to upstream. It is identical to your master, except I rebased it on upstream/master to include the two commits you are currently behind. Binaries should appear here: https://beta.rclone.org/branch/pasnox-symlink/

pasnox commented:
> I pushed a branch [pasnox-symlink](https://github.com/rclone/rclone/tree/pasnox-symlink) to upstream. It is identical to your master, except I rebased it on upstream/master to include the two commits you are currently behind. Binaries should appear here: https://beta.rclone.org/branch/pasnox-symlink/

Awesome, thank you very much !

@ncw Any ETA for a review ? or should somebody else do the review ?

pasnox commented:
> I pushed a branch [pasnox-symlink](https://github.com/rclone/rclone/tree/pasnox-symlink) to upstream. It is identical to your master, except I rebased it on upstream/master to include the two commits you are currently behind. Binaries should appear here: https://beta.rclone.org/branch/pasnox-symlink/

I rebased against master and cherry picked not yet merged @ncw s3 use server modtime.
Would you please update the branch you create for me ? so that I can test in good condition on my target devices ?

Thanks,

albertony commented:
Pushed it. Fresh binaries should be ready in some minutes.

xmready commented:
Is this getting close to being merged?

pasnox commented:
> Is this getting close to being merged?

I'm testing it successfully since some weeks now and it works fine. Now waiting for review, so it depends how fast they test and review it as well

pasnox commented:
@ncw Hey, are you planning to make a review soon ?
This is a major feature that could go out of sync quite easily, and I would like to avoid the need to rebase and/or fix stuff regularly until it's reviewed / merged as it's time consuming for me, and knowledge on the base code reduce rapidly.

xmready commented:
@ncw this pull request is extremely desirable. Any possibility you'll have time to review it?

ncw commented on vfs/file.go:
```diff
@@ -89,15 +96,21 @@ func (f *File) IsDir() bool {
-	mode = f.d.vfs.Opt.FilePerms
-	if f.appendMode {
-		mode |= os.ModeAppend
+	// We could return the passed *create* FileMode, but it seems rclone want *always* Opt.FilePerms
```
This doesn't look right - maybe the tests need fixing?

pasnox replied:
No idea :|
It is hard for me to understand all correctly, there is lots of code in the VFS, so i have no idea how you imagined /designated the api initially, so that was more a question for you I added this comment :)

pasnox replied:
Closing as part of the old trashed code

ncw commented on vfs/file.go:
```diff
@@ -466,7 +479,17 @@ func (f *File) setObject(o fs.Object) {
 	f.mu.Lock()
+	oldMode := f.mode
 	f.o = o
+	// FIXME: Is this needed at all ?
```
I would have hoped that this isn't needed. It justs writesback the rclone object into the VFS object - it shouldn't change the handling of the VFS object.

pasnox replied:
Ok, i will make some testing removing this code

pasnox replied:
Tests still pass without that, so reverted.

ncw commented on vfs/file_test.go:
```diff
@@ -1,13 +1,20 @@
 	"io/ioutil"
+	"log"
 	"os"
+	_ "path"
```
I don't think you need `_ "path"` and `"path"`??? Same with `"string"`

ncw commented on vfs/file_test.go:
```diff
@@ -17,20 +24,74 @@ import (
 	require.True(t, node.Mode().IsRegular())
 
-	return r, vfs, node.(*File), file1, cleanup
+	// For some reason this does not works - Stat fails and return file not found ?
```
???

pasnox replied:
Yep, never had the code commented working, if you have an idea ?
I can also just remove it as well entirely, but I would prefer to understand why it did not worked.
Initially this function was expecting to create a regular file and a symlink file, but for some reason trying to add more than one file failed and I did not found why.

pasnox replied:
Closing as part of the old trashed code

ncw commented on vfs/help.go:
```diff
@@ -258,6 +258,7 @@ read of the modification time takes a transaction.
     --no-modtime      Don't read/write the modification time (can speed things up).
     --no-seek         Don't allow seeking in files.
     --read-only       Only allow read-only access.
+    --links           Translate symlinks to/from regular files with a '.rclonelink' extension.
```
There is already a local backend flag called `--links`. I think this might need to be `--vfs-links`.

pasnox replied:
Closing, will focus on the next comment instead.

ncw commented on vfs/vfsflags/vfsflags.go:
```diff
@@ -24,6 +25,7 @@ func AddFlags(flagSet *pflag.FlagSet) {
 	flags.DurationVarP(flagSet, &Opt.DirCacheTime, "dir-cache-time", "", Opt.DirCacheTime, "Time to cache directory entries for")
 	flags.DurationVarP(flagSet, &Opt.PollInterval, "poll-interval", "", Opt.PollInterval, "Time to wait between polling for changes, must be smaller than dir-cache-time and only on supported remotes (set 0 to disable)")
 	flags.BoolVarP(flagSet, &Opt.ReadOnly, "read-only", "", Opt.ReadOnly, "Only allow read-only access")
+	flags.BoolVarP(flagSet, &Opt.Links, "links", "l", Opt.Links, "Translate symlinks to/from regular files with a '"+fs.LinkSuffix+"' extension")
```
These clash with the local backend flags - I thought pflag would throw an error here at runtime?

pasnox replied:
Closing as part of the old trashed code

ncw reviewed (commented):
Thank you for your hard work on this, and apologies for not reviewing sooner.

I've put comments inline.

The main thing missing from this is tests. The VFS is very complicated and the only way we keep from breaking it is to have lots of tests.

This needs more tests in

- `vfs/dir_test.go` - As there are new paths in there which aren't being tested.
- `vfs/vfs_test.go` - to test the additions to vfs.go
- `vfs/vfstest/` - These are integration tests and they run in the vfs tests **and** in the mount tests - this is how you test the features added to cmd/mount cmd/cmount cmd/mount2

You might have to make sure the symlink tests don't run on Windows - not really sure whether they will work there or not.

When you are done you should be able to run `bin/test-repeat-vfs.sh 1` with no errors (this runs all the mount and vfs tests together and is useful for testing VFS changes). You can then try a soak test with `bin/test-repeat-vfs.sh 10`. You'll have to run the tests for cmd/mount and cmd/mount2 manually as these aren't part of the VFS soak test as they aren't 100% reliable!

Can you remove fix all the FIXMEs and notes?

Your commits tell a nice story which is good. Can you make sure all commits are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages)

eg `mount,cmount,mount2: Add full support for symlinks in vfs` The change release generator knows to split this up into 3 changelog lines one for each mount style.

Can you rebase on master and fix the conflict.

I'd like to get this merged for 1.61 if at all possible. That will be towards the end of December.

Thank you :-)

pasnox commented:
> Thank you for your hard work on this, and apologies for not reviewing sooner.

No problems

> The main thing missing from this is tests. The VFS is very complicated and the only way we keep from breaking it is to have lots of tests.
> * `vfs/dir_test.go` - As there are new paths in there which aren't being tested.
> * `vfs/vfs_test.go` - to test the additions to vfs.go
> * `vfs/vfstest/` - These are integration tests and they run in the vfs tests **and** in the mount tests - this is how you test the features added to cmd/mount cmd/cmount cmd/mount2

Hm, I don't understand, I do already provided tests, especially in vfs/file_test and vfs/dir_test, could you be more explicit of what is missing / untested correctly ? Honestly having help here would be helpful.

> 
> You might have to make sure the symlink tests don't run on Windows - not really sure whether they will work there or not.

From the docs I read (can't remember where), symlinks should be supported on windows with fuse.
I also picked / pasted some windows specific docs from fuse when documenting symlink functions in the vfs.
At least the tests pass on your CI.

> When you are done you should be able to run `bin/test-repeat-vfs.sh 1` with no errors (this runs all the mount and vfs tests together and is useful for testing VFS changes). You can then try a soak test with `bin/test-repeat-vfs.sh 10`. You'll have to run the tests for cmd/mount and cmd/mount2 manually as these aren't part of the VFS soak test as they aren't 100% reliable!

How would I triggers tests for cmd/mount / cmd/mount2 ?
Currently, I do run that to make my tests pass locally:
`cd ${HOME}/Projects/rclone/vfs && go test -tags cmount -v`
Is this not enough to run all vfs tests ?

> 
> Can you remove fix all the FIXMEs and notes?

Sure

> Your commits tell a nice story which is good. Can you make sure all commits are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages)

Will try

> Can you rebase on master and fix the conflict.

Done

> I'd like to get this merged for 1.61 if at all possible. That will be towards the end of December.

Awesome, I hope we can make it.

pasnox commented on cmd/cmount/fs.go:
```diff
@@ -330,13 +324,15 @@ func (fsys *FS) CreateEx(filePath string, mode uint32, fi *fuse.FileInfo_t) (err
 	}
-	file, err := parentDir.Create(leaf, fi.Flags)
+	// translate the fuse flags to os flags
+	osFlags := translateOpenFlags(fi.Flags) | os.O_CREATE
```
The previous code was already altering the flags later, i just move it atop so that he same flags get passed to create and open, for consistency.
The flags are compatible, that's why they worked, but having them translated looked the correct way.

pasnox replied:
Closing as part of the old trashed code

pasnox commented on cmd/mount/file.go:
```diff
@@ -32,7 +33,7 @@ func (f *File) Attr(ctx context.Context, a *fuse.Attr) (err error) {
 	a.Gid = f.VFS().Opt.GID
 	a.Uid = f.VFS().Opt.UID
-	a.Mode = f.VFS().Opt.FilePerms
+	a.Mode = f.File.Mode() &^ os.ModeAppend
```
Before my changes in vfs File struct, you were holding separately the permissions and the append mode.
This is now unified in an os.FileMode (mode), so as you were never returning this append flag before, I ensured the old behavior was respected.

pasnox replied:
Hm, just spotted that:

https://github.com/rclone/rclone/pull/6337/commits/f88e6112921a09e93f3fe403facb9340bd83f496#diff-b83d7214cb08180c46ed256e5b1dd9017fb7ae1eb36c30ca65f4e30cc498cc91L102

So it seems the vfs would forward the append mode, so should I keep it here and consider the old code a bug ?

pasnox replied:
Closing as part of the old trashed code

ncw commented:
> > The main thing missing from this is tests. The VFS is very complicated and the only way we keep from breaking it is to have lots of tests.
> > 
> > * `vfs/dir_test.go` - As there are new paths in there which aren't being tested.
> > * `vfs/vfs_test.go` - to test the additions to vfs.go
> > * `vfs/vfstest/` - These are integration tests and they run in the vfs tests **and** in the mount tests - this is how you test the features added to cmd/mount cmd/cmount cmd/mount2
> 
> Hm, I don't understand, I do already provided tests, especially in vfs/file_test and vfs/dir_test, could you be more explicit of what is missing / untested correctly ? Honestly having help here would be helpful.

Sure.

What I do is use the [go coverage tool](https://go.dev/blog/cover) by adding the flags `-cover -coverprofile cover.prof` to the `go test` command. I then use my editor (which understands the `cover.prof` files) to check that all my additions are covered by the tests. The link above shows how to do it without editor support.

Ideally everything you've added (except for unlikely error paths) should be covered with your new tests.

This will find anything you've missed in  `vfs/*.go`.

The integration tests in `vfs/vfstest/` are different. These are designed to exercise the VFS as a blackbox test and you need to try to cover all your new code with these tests too. These tests are run on the VFS directly, and on a mounted file system when you run the tests in `cmd/[c]mount` so its really important they cover the new features too.

So, for example, I'd like to see a test in here

https://github.com/rclone/rclone/blob/7672cde4f348c6485c7de39795ef7d538737b46f/vfs/vfstest/file.go#L1

Which creates a symlink, lists it, reads content through it, renames it and deletes it.

> > You might have to make sure the symlink tests don't run on Windows - not really sure whether they will work there or not.
> 
> From the docs I read (can't remember where), symlinks should be supported on windows with fuse. I also picked / pasted some windows specific docs from fuse when documenting symlink functions in the vfs. At least the tests pass on your CI.

I think we'll have to get the tests running there and see what happens!

> > When you are done you should be able to run `bin/test-repeat-vfs.sh 1` with no errors (this runs all the mount and vfs tests together and is useful for testing VFS changes). You can then try a soak test with `bin/test-repeat-vfs.sh 10`. You'll have to run the tests for cmd/mount and cmd/mount2 manually as these aren't part of the VFS soak test as they aren't 100% reliable!
> 
> How would I triggers tests for cmd/mount / cmd/mount2 ? Currently, I do run that to make my tests pass locally: `cd ${HOME}/Projects/rclone/vfs && go test -tags cmount -v` Is this not enough to run all vfs tests ?

This only runs the VFS tests, you need to run the mount tests too.

So

```
cd ${HOME}/Projects/rclone/vfs
# Run the VFS tests in all subdirectories too
go test -tags cmount -v ./...
# This test uses the tests in vfs/vfstest
cd ${HOME}/Projects/rclone/mount/cmount
go test -tags cmount -v
# This test uses the tests in vfs/vfstest
cd ${HOME}/Projects/rclone/mount/mount
go test -v
# This test uses the tests in vfs/vfstest
cd ${HOME}/Projects/rclone/mount/mount2
go test  -v
```

> > Can you remove fix all the FIXMEs and notes?
> 
> Sure
> 
> > Your commits tell a nice story which is good. Can you make sure all commits are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages)
> 
> Will try

Can you get the first line so they have the component first, eg

> VFS.Dir.stat: Add support for symlinks

Should be something like

> vfs: VFS.Dir.stat: Add support for symlinks

This means the changelog generator will group all the changes under `vfs` instead of `VFS.Dir.stat`.

If you look at the categories in [the changelog](https://rclone.org/changelog/) you'll see what I'm after. Put `vfs:` in lower to match the file system, the changelog generator capitalizes it.

> > I'd like to get this merged for 1.61 if at all possible. That will be towards the end of December.
> 
> Awesome, I hope we can make it.

I think we will :-)

xmready commented:
Hope this gets merged for 1.61  !!

pasnox commented:
@ncw I moved the old implementation in my fork in the master-symlinks-v1 branch.
I pushed here the new refactoring (call it v2), which fix flaws encountered, especially in the VFS code path.
This comes with full unit tests.

I closed all old opened comments as the code is quite different now, please review again and add comments if needed.

ncw commented:
Thanks for doing this. I hope to have a bit of time to review this after the 1.61 release is out (sorry I've got too much to do to get this in too).

pasnox commented:
@ncw CI Looks great now, just minors lints to fix. Still I have no idea ho to fix them? I did introduce no new types in the pointed code?

ncw commented:
Looks like you have a few files to gofmt that's all.

pasnox commented:
@albertony @ncw Can one of you push those changes in the pasnox branch so we can generate binaries that can be tested?

albertony commented:
> Can one of you push those changes in the pasnox branch so we can generate binaries that can be tested?

✔️ Done

New betas should in a few minutes appear in the same place as before: https://beta.rclone.org/branch/pasnox-symlink/

xmready commented:
hows the testing coming?

pasnox commented:
> hows the testing coming?

Hm I thought I had installed latest build and testing since 1month... But it's not 😭 So I just installed it right now, so far works fine but I will have to test a bit longer. And anybody can test, build are publicly accessible.

pasnox commented:
> > hows the testing coming?
> 
> Hm I thought I had installed latest build and testing since 1month... But it's not sob So I just installed it right now, so far works fine but I will have to test a bit longer. And anybody can test, build are publicly accessible.

Tested it for long enough now, nothing to report apart it works fine and reliably on Linux.
I do expect macOS follow well, even though I dd not tested it (as a posix compliant OS).
So far windows works partially, but issues remains. Still windows is not a priority for me (I do even have no windows machine at hands). So testing and reporting issues could help those being fixed.

@ncw what do we do now ?

xmready commented:
@ncw paging Dr. Nick

pasnox commented:
@albertony @ncw Hey, I just pushed fixes for working symlinks support on Windows. Could you please trigger a CI builds and provide url to grab resulting binaries ?
I also enabled the unit tests for the windows platform, if all goes fine then we could consider those changes mature enough to be reviewed carefully and merged !

ncw commented:
Sorry I'm running behind with PRs. I'll be able to pay this some attention when we've got 1.63 out the door :-)

ncw commented:
I've rebased this on master and pushed it here to make some test binaries

[v1.64.0-beta.7304.2a00583f5.pr-6337-mount-symlink](https://beta.rclone.org/branch/pr-6337-mount-symlink/v1.64.0-beta.7304.2a00583f5.pr-6337-mount-symlink/) on branch [pr-6337-mount-symlink](https://github.com/rclone/rclone/tree/pr-6337-mount-symlink) (uploaded in 15-30 mins)

You might want to pull that branch and use it @pasnox since the rebase wasn't trivial! But I didn't want to push to your PR without your permission - I can do that if you want.

pasnox commented:
> I've rebased this on master and pushed it here to make some test binaries
> 
> [v1.64.0-beta.7304.2a00583f5.pr-6337-mount-symlink](https://beta.rclone.org/branch/pr-6337-mount-symlink/v1.64.0-beta.7304.2a00583f5.pr-6337-mount-symlink/) on branch [pr-6337-mount-symlink](https://github.com/rclone/rclone/tree/pr-6337-mount-symlink) (uploaded in 15-30 mins)
> 
> You might want to pull that branch and use it @pasnox since the rebase wasn't trivial! But I didn't want to push to your PR without your permission - I can do that if you want.

@ncw Thank you. I also rebased my copy upon master, and after diffing your changes it's identical. So all fine.
I would really want we advance here, so we have to decide things:
1. We disable windows support and postpone it for a next version so that macos / linux can have this feature shortly. I do use symlinks since a while now (more than a year), And I did not encountered any issue (apart possible slowdown when using thousands of symlinks in the same folder, but i suppose it's something you could easily fix by caching the link target in the metadata maybe in a follow up).
2. You help us fixing the remaining issue for windows which is really hard for me to fix has I don't know really well all the internals of rclone, especially workaround this annoying winfsp symlink hack. You can see the latest WIP commit that is a start but clearly not great and not working in all cache mode. It was just a try to confirm the needed work around.

Can we have your support here so we finally merge that feature, I'm really busy with side things and I don't have much time anymore to maintains the changes :|

xmready commented:
paging doctor @ncw

eharris commented:
I hope this gets merged soon, seems like a great feature

hongkongkiwi commented:
Let's get this merged and disable Windows support? It would be such a disappointment to have all this heroic amount of work go to waste. I'm so excited for this feature (on Linux)!

xmready commented:
I've been cheering on this PR for almost two years now. Would love nothing more than to see it succeed!

FurloSK commented:
hi, this is a great addition! One question: will this change allow Google shortcuts to be viewed in Rclone as aliases?

xmready commented:
@ncw v1.67 has come and gone, yet this PR is still open. Can we ever expect this amazing feature to make it into Rclone?

peter-tar commented:
I would love to see this merged as well.

ncw commented:
I have rebased this PR. I'm going to go through it to see what needs to be done to it to get it merged.

pasnox commented:
Thanks for your interest again in this area @ncw .
As far as I remember, the implementation is fully correct and working on all paltforms.
The only issue is the windows version using the winfsp or something named like for for the fuse implementation which is doing dirty hack to support symlink causing rclone issue. I cna't remember where, but I explained the problem very well.
I can't wait to have it merged, finally ! When, if, done, pelase ping me afterwards as tehre is some things we could discuss to make better performances with the symlink lookups, but each things at its time.

ncw commented:
Thank you for your work on this @pasnox 

I'm going to merge this into a new branch fix-2975-mount-symlinks as I'm going to work on it a bit before merging to master.

In particular I'm going to push **all** the symlink logic into the vfs layer so there is no duplicated code in the mount/cmount/mount2.

I'm haven't got my head around the windows support yet though.

I'll merge this, then I'll push an updated version and post a beta for people to try.

ncw commented:
Here is a beta of the current build if anyone wants to give it a try

[v1.69.0-beta.8381.204b90a79.fix-2975-mount-symlinks](https://beta.rclone.org/branch/fix-2975-mount-symlinks/v1.69.0-beta.8381.204b90a79.fix-2975-mount-symlinks/) on branch [fix-2975-mount-symlinks](https://github.com/rclone/rclone/tree/fix-2975-mount-symlinks) (uploaded in 15-30 mins)

ncw commented:
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.69

Thank you @pasnox 

Any more testing gratefully received :-)

pasnox commented:
Great news @ncw !
I'm glad this finally landed in master!
I will test it, is there a rpi5 64bits build i can grab somewhere??

ncw commented:
@pasnox wrote
> Great news! I'm glad this finally landed in master! I will test it

Thanks

> is there a rpi5 64bits build i can grab somewhere??

Sure - this one should work https://beta.rclone.org/rclone-beta-latest-linux-arm.zip (or .deb or .rpm if you prefer)

> Were you able to fix the win FS issue I had ?

I dedicated to get this merged. I think there are likely still problems under Windows but that is a fix for another day :-)

pasnox commented:
Please delete pr-6337-mount-symlink branch, it's no more needed then.
Yeah, let tackle windows issues as they comes, maybe winfsp is not even used by rclone users ;)
Do you have performances benchmark for symlink support ?
Improving symlinks performance was somehting i wanted to work on / fix at a later step, but maybe your refactoring fix it entirely. Basically, evaluating symlinks would trigger network request just to read its target content. It's maybe something we would prefer to store locally, maybe in the metadata so that's it's fast and does not require extra network request anytime we require evaluating symlinks target.

ncw commented:
> Please delete pr-6337-mount-symlink / pasnox-symlinks branches, it's no more needed then. 

Done

> Yeah, let tackle windows issues as they comes, maybe winfsp is not even used by rclone users ;)

Lots of people use winfsp but I think symlinks are much less used on Windows.

> Do you have performances benchmark for symlink support ? Improving symlinks performance was somehting i wanted to work on / fix at a later step, but maybe your refactoring fix it entirely. Basically, evaluating symlinks would trigger network request just to read its target content. It's maybe something we would prefer to store locally, maybe in the metadata so that's it's fast and does not require extra network request anytime we require evaluating symlinks target.

The symlinks can get stored in the VFS cache so using `--vfs-cache-mode full` will cache them.

The OS should cache them also so they shouldn't get read too often, but I didn't measure performance, I was more worried about correctness for this pass.

pasnox commented:
I tested over a sftp remote (no cache) and it worked fine.
I was able to:
- create files/dirs
- create files/dirs links pointing to the previous created entries
- move files/links in between
- delete files / links / dirs
So far for me, it seems to works great.
One exception, you did not fully refactored / ported corner case #3 from tests, as such, we are able to copy/move files with same name into dirs already containing the same name. I would really want you fix it before releasing the next version.
See https://github.com/rclone/rclone/commit/f1d2f2b2c85599553a92eb4b2e5d0e7134ff8edb#diff-4b3637c29d010b5284e9bc43437f32ffb7a5a6cb85c0bea56f4bbb7df23c09a4R237

ncw commented:
Thanks for testing @pasnox much appreciated 

I think those corner case tests got lost in the code reshuffle, alas 

Can you make an issue about the problem please so we don't forget - thank you .

Do you want to have a go at fixing?

pasnox commented:
> Can you make an issue about the problem please so we don't forget - thank you .
https://github.com/rclone/rclone/issues/8245

> Do you want to have a go at fixing?
I will have no time for that sorry, please go ahead, most likely reading the original PR code should show how exactly it's to be checked.
